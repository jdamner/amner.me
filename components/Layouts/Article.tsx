"use client";

import React from "react";
import ButtonLink from "../Atoms/Links/ButtonLink";
import Container from "./Container";
import TwoThirds from "./TwoThirds";
import Title from "../Atoms/Title";

export default function Article({
  title,
  children,
}: {
  children?: React.ReactNode;
  title?: React.ReactNode;
}) {
  return (
    <Container>
      <div className="p-5">
        <TwoThirds
          first={
            title && (
              <div className="md:sticky md:top-5 ">
                <Title>{title}</Title>
                <span className="block">
                  <ButtonLink href="/">Â« Home</ButtonLink>
                </span>
              </div>
            )
          }
        >
          {children}
        </TwoThirds>
      </div>
    </Container>
  );
}
