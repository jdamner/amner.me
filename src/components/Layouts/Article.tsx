import React from "react";
import { WithChildren } from "../../types/WithChildren.type";
import ButtonLink from "../Links/ButtonLink";
import Container from "./Container";
import TwoThirds from "./TwoThirds";
export default function Article({ first, image, children, offset = false }: WithChildren & { first?: React.ReactNode | React.ReactNode[], image?: React.ReactNode | React.ReactNode[], offset?: boolean }): React.JSX.Element {
    return (
        <Container>
            <TwoThirds top first={
                (first||image) && <div className={offset ? '-mt-10 md:-mt-20 h-full' : ''}>
                    {image}
                    <div className='md:sticky md:top-5'>
                        {first}
                        <span className='hidden md:block'>
                            <ButtonLink href='/'>Â« Home</ButtonLink>
                        </span>
                    </div>
                </div>
            }>
                {children}
            </TwoThirds>
        </Container>
    )
}