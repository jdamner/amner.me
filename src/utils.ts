/**
 * Helper functions
 */
import fs from 'fs'
import matter from 'gray-matter'
import { join } from 'path'
import { slug } from 'github-slugger'

import type { ReactNode } from 'react'
import type { MdFile } from './types'

/**
 * Makes an object JSON parseable
 * 
 * @param {any} obj The object to be parsed
 * @returns {any}
 */
export function makeJsonParseable<Type>(obj: Type): Type {
	return JSON.parse(JSON.stringify(obj)) as Type
}

/**
 * Reads a markdown file
 */
export async function readMdFile(filename: string, directory: string = 'content/'): Promise<MdFile> {
	const path = `${directory}/${filename}`
	const fullPath = `${process.cwd()}/${path}`
	const fileContents = fs.readFileSync(fullPath, 'utf8')
	return {
		slug: slug(filename.replace(/\.md$/, '')),
		...matter(fileContents),
	}
}

/**
 * Reads all slugs from a directory
 * 
 * @param {string} directory The directory to read from
 * 
 * @returns {string[]}
 */
export function getFilesInDir(directory: string): string[] {
	const fullPath = join(process.cwd(), directory)
	return fs.readdirSync(fullPath)
}

/**
 * Converts all headings in a markdown file to have IDs 
 * Converts 
 * ## Heading 1 
 * to become
 * ## Heading 1{#heading-1}
 * 
 * @param {string} markdown The markdown to convert
 * 
 * @returns {string}
 */
export function convertHeadingsToAnchorLinks(markdown: string): string {
	return markdown.replace(/(#{1,6})\s(.+)\n/gm, (match, p1, p2) => {
		return `${p1} ${p2}{#${slug(p2)}}\n`
	})
}


/**
 * Converts a react node to a slug
 * 
 * Usually the value is a string, but sometimes it contains
 * a `<code>` element, which is a react node. This function
 * converts the value to a string.
 */
export function makeSlug(value: ReactNode | ReactNode[]): string {
	if (! Array.isArray(value)) return slug(value?.toString() ?? '');
	return slug(
		value.map((node) => {
			if (typeof node === 'string' ||
				typeof node === 'number' ||
				typeof node === 'boolean'
			) return node;

			if (typeof node === 'object' && node !== null && 'props' in node) {
				return makeSlug(node.props.children)
			}

			return ''
		}).join('')
	)
}