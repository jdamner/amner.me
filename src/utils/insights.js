import { track, parameters, init } from 'insights-js';

const pageView = () => {
    event('page-view');
}

const event = (eventName, eventProperties = {}) => {
    init('HbSIygPWVvMzXFRP');
    track({
        id: eventName,
        parameters: {
            ...parameters,
            ...eventProperties,
        },
    });
}

export { pageView, event };