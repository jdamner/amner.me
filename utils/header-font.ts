import { Bebas_Neue as Font } from "next/font/google";
export const font = Font({
  weight: "400",
  subsets: ["latin", "latin-ext"],
});
