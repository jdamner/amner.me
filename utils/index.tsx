/* External Deps */
import { readFileSync, readdirSync } from "fs";
import { resolve } from "path";
import matter from "gray-matter";
import { slug } from "github-slugger";

/* Components */
import { Markdown } from "@/components/Markdown";

/* Constants */
const postsDirectory: string = resolve("content/blog/");
const servicesDirectory: string = resolve("content/services");
const cvDirectory: string = resolve("content/cv");
const projectsDirectory: string = resolve(cvDirectory, "projects");
const referencesDirectory: string = resolve(cvDirectory, "references");
const employmentDirectory: string = resolve(cvDirectory, "employment");
const educationDirectory: string = resolve(cvDirectory, "education");

type MatterData = ReturnType<typeof matter>["data"];
type AttributeValidator<T extends MatterData> = (data: MatterData) => data is T;

const defaultAttributeValidator = <T extends MatterData>(
  data: MatterData,
): data is T => {
  return typeof data === "object" && data !== null;
};

/**
 * Gets all files in a directory
 */
const getAllFrom = <T extends MatterData>(
  directory: string,
  attributeValidator: AttributeValidator<T> = defaultAttributeValidator,
) =>
  readdirSync(directory).map((filename) =>
    readMdFile<T>(filename, directory, attributeValidator),
  );

/**
 * Reads a markdown file
 */
const readMdFile = <T extends MatterData>(
  filename: string,
  directory: string = "content/",
  attributeValidator: AttributeValidator<T> = defaultAttributeValidator,
) =>
  parseMarkdown<T>(
    readFileSync(`${directory}/${filename}`, "utf8"),
    filename,
    attributeValidator,
  );

/**
 * Parses markdown file contents
 */
export const parseMarkdown = <T extends MatterData>(
  fileContents: string,
  filename: string,
  attributeValidator: AttributeValidator<T> = defaultAttributeValidator,
) => {
  const { content, data } = matter(fileContents);

  if (!attributeValidator(data)) {
    console.log(data);
    throw new Error(`Invalid attributes in ${filename}` + JSON.stringify(data));
  }

  return {
    data,
    slug: slug(filename.replace(/\.mdx$/, "")),
    content: <Markdown>{content}</Markdown>,
  };
};

/**
 * Gets all Posts
 */
const getAllPosts = () =>
  getAllFrom<WithTitle & WithThumbnail & WithDate>(
    postsDirectory,
    (data): data is WithTitle & WithThumbnail & WithDate =>
      hasTitle(data) && hasThumbnail(data) && hasDate(data),
  );

export const useAllServices = () =>
  getAllFrom<WithTitle>(servicesDirectory, hasTitle);

export const getAllProjects = () =>
  getAllFrom<WithTitle & WithLink>(
    projectsDirectory,
    (data): data is WithTitle & WithLink => hasTitle(data) && hasLink(data),
  );

export const useAllReferences = () =>
  getAllFrom<WithTitle & WithWebsite & WithPhone & WithEmail>(
    referencesDirectory,
    (data): data is WithTitle & WithWebsite & WithPhone & WithEmail =>
      hasTitle(data) && hasWebsite(data) && hasPhone(data) && hasEmail(data),
  );

export const useAllEmployment = () =>
  getAllFrom<
    WithTitle & WithSubtitle & WithDateString & WithRole & WithWebsite
  >(
    employmentDirectory,
    (
      data,
    ): data is WithTitle &
      WithSubtitle &
      WithDateString &
      WithRole &
      WithWebsite =>
      hasTitle(data) &&
      hasSubtitle(data) &&
      hasDateString(data) &&
      hasRole(data) &&
      hasWebsite(data),
  );

export const useAllEducation = () =>
  getAllFrom<WithTitle & WithDateString>(
    educationDirectory,
    (data): data is WithTitle & WithDateString =>
      hasTitle(data) && hasDateString(data),
  );

export const getPost = (slug: string) =>
  getAllPosts().find((post) => post.slug === slug);

export const getAllPostLinks = () =>
  getAllPosts()
    .map((post) => ({
      title: post.data.title,
      thumbnail: post.data.thumbnail,
      slug: post.slug,
      date: post.data.date,
    }))
    .sort((a, b) => (a.date > b.date ? -1 : 1));

type WithTitle = { title: string };
type WithSubtitle = { subtitle: string };
type WithLink = { link: string };
type WithThumbnail = { thumbnail: string };
type WithDate = { date: Date };
type WithDateString = { date: string };
type WithPhone = { phone: string };
type WithWebsite = { website: string };
type WithEmail = { email: string };
type WithRole = { role: string };

const hasTitle = (data: MatterData): data is WithTitle =>
  defaultAttributeValidator(data) &&
  "title" in data &&
  typeof data.title === "string";

const hasSubtitle = (data: MatterData): data is WithSubtitle =>
  "subtitle" in data && typeof data.subtitle === "string";

const hasLink = (data: MatterData): data is WithLink =>
  "link" in data && typeof data.link === "string";

const hasThumbnail = (data: MatterData): data is WithThumbnail =>
  "thumbnail" in data && typeof data.thumbnail === "string";

const hasDate = (data: MatterData): data is WithDate =>
  "date" in data && data.date instanceof Date;

const hasDateString = (data: MatterData): data is WithDateString =>
  "date" in data && typeof data.date === "string";

const hasPhone = (data: MatterData): data is WithPhone =>
  "phone" in data && typeof data.phone === "string";

const hasWebsite = (data: MatterData): data is WithWebsite =>
  "website" in data && typeof data.website === "string";

const hasEmail = (data: MatterData): data is WithEmail =>
  "email" in data && typeof data.email === "string";

export const hasRole = (data: MatterData): data is WithRole =>
  "role" in data && typeof data.role === "string";
