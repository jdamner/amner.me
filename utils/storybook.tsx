import { faker } from "@faker-js/faker";
import { MdFile, PostLinkType } from "../types";
import matter from "gray-matter";
import React from "react";

type MdAttributes = Record<string, string>;

faker.seed(123);

export const generateMdFile = <T extends MdAttributes>(
  attributes: T,
  content?: string,
): Omit<MdFile<string>, "data"> & { data: T } => {
  return matter(generateMD(attributes, content)) as Omit<
    MdFile<string>,
    "data"
  > & { data: T };
};

export const generateMdContent = (): string => {
  return (
    `# ${faker.lorem.words(3)}\n` +
    Array.from({ length: 3 }, () => {
      return (
        `## ${faker.lorem.words(2)}\n` +
        faker.lorem.paragraph() +
        "\n" +
        Array.from({ length: 3 }, () => {
          return (
            `### ${faker.lorem.words(2)}\n` + `${faker.lorem.paragraph()}\n`
          );
        }).join("")
      );
    }).join("")
  );
};

export const generatePostLink = (): PostLinkType => {
  return {
    title: faker.lorem.words(3),
    thumbnail: faker.image.url(),
    slug: faker.lorem.slug(),
  };
};

export const generatePostLinks = (count: number): PostLinkType[] => {
  return Array.from({ length: count }, () => generatePostLink());
};

export const generateMD = (
  attributes: MdAttributes,
  content?: string,
): string => {
  const frontMatter = generateMDFrontMatter(attributes);
  return `${frontMatter}\n${content ?? generateMdContent()}`;
};

export const generateMDFrontMatter = (attributes: MdAttributes): string => {
  return (
    "---\n" +
    Object.keys(attributes)
      .map((key) => `${key}: ${attributes[key]}`)
      .join("\n") +
    "\n---"
  );
};

export const generatePlaceholderContent = (
  size: number = 5,
): React.ReactNode => {
  return (
    <>
      {Array.from({ length: size }).map((_, i) => (
        <p key={i}>{faker.lorem.paragraph(i + 1)}</p>
      ))}
    </>
  );
};
