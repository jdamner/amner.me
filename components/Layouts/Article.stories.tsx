import type { Meta, StoryObj } from "@storybook/nextjs";

import { faker } from "@faker-js/faker";
import Article from "./Article";
import { generatePlaceholderContent } from "../../utils/storybook";

type Story = StoryObj<typeof Article>;

export default {
  title: "Layout/Article",
  component: Article,
} as Meta;

faker.seed(123);

export const Default: Story = {
  args: {
    children: generatePlaceholderContent(),
  },
};
