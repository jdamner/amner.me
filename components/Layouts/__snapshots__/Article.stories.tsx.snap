// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Layout/Article Default smoke-test 1`] = `
<div class="bg-slate-100 dark:bg-slate-900 text-slate-900 dark:text-slate-300">
  <div class="container mx-auto py-5 px-3">
    <div class="flex flex-col md:flex-row mb-5 gap-5 md:items-stretch">
      <div class="basis-1/3">
        <div class>
          <img alt
               width="500"
               height="500"
               decoding="async"
               data-nimg="1"
               class="border-2 border-slate-500 mb-5 bg-white"
               srcset="https://picsum.photos/seed/hqeGyIIuq/640/480?w=640&amp;q=75 1x, https://picsum.photos/seed/hqeGyIIuq/640/480?w=1080&amp;q=75 2x"
               src="https://picsum.photos/seed/hqeGyIIuq/640/480?w=1080&amp;q=75"
               style="color: transparent;"
          >
          <div class="md:sticky md:top-5">
            <h1>
              Title
            </h1>
            <span class="hidden md:block">
              <a class="border-2 border-slate-500 uppercase font-bold bg-white dark:bg-slate-900 px-4 py-2 text-slate-800 dark:text-slate-200 hover:bg-slate-100 dark:hover:bg-slate-800 hover:text-black dark:hover:text-white"
                 href="/"
              >
                « Home
              </a>
            </span>
          </div>
        </div>
      </div>
      <div class="w-full">
        <p>
          Vomer cunctatio stillicidium nisi denique apud cruciamentum cui conicio.
        </p>
        <p>
          Suspendo commodo decens caries adulatio sub cubicularis paens. Ratione auctor decerno atque agnosco exercitationem summa exercitationem.
        </p>
        <p>
          Reiciendis coadunatio ulciscor deprimo surculus taedium. Contabesco suppono cur commodi sollers conturbo corrumpo. Canonicus claudeo voveo quos.
        </p>
        <p>
          Aliquam tergiversatio damno acsi curvo atavus depulso arguo cursim. Color valetudo cursim. Varietas suadeo vinco doloremque desipio delectus quia minus ancilla. Comes derideo cuppedia statua usitas.
        </p>
        <p>
          Casso amitto denuncio deputo voro candidus earum debilito placeat tantillus. Denique trepide thema. Cimentarius impedit damnatio conicio absque cognomen confugo. Strenuus stabilis ustulo utique ceno dolore. Sortitus censura occaecati tui quibusdam cerno.
        </p>
      </div>
    </div>
  </div>
</div>
`;

exports[`Layout/Article NoChildren smoke-test 1`] = `
<div class="bg-slate-100 dark:bg-slate-900 text-slate-900 dark:text-slate-300">
  <div class="container mx-auto py-5 px-3">
    <div class="flex flex-col md:flex-row mb-5 gap-5 md:items-stretch">
      <div class="basis-1/3">
        <div class>
          <img alt
               width="500"
               height="500"
               decoding="async"
               data-nimg="1"
               class="border-2 border-slate-500 mb-5 bg-white"
               srcset="https://picsum.photos/seed/0EysVzAqch/640/480?w=640&amp;q=75 1x, https://picsum.photos/seed/0EysVzAqch/640/480?w=1080&amp;q=75 2x"
               src="https://picsum.photos/seed/0EysVzAqch/640/480?w=1080&amp;q=75"
               style="color: transparent;"
          >
          <div class="md:sticky md:top-5">
            <h1>
              Title
            </h1>
            <span class="hidden md:block">
              <a class="border-2 border-slate-500 uppercase font-bold bg-white dark:bg-slate-900 px-4 py-2 text-slate-800 dark:text-slate-200 hover:bg-slate-100 dark:hover:bg-slate-800 hover:text-black dark:hover:text-white"
                 href="/"
              >
                « Home
              </a>
            </span>
          </div>
        </div>
      </div>
      <div class="w-full">
      </div>
    </div>
  </div>
</div>
`;

exports[`Layout/Article NoImage smoke-test 1`] = `
<div class="bg-slate-100 dark:bg-slate-900 text-slate-900 dark:text-slate-300">
  <div class="container mx-auto py-5 px-3">
    <div class="flex flex-col md:flex-row mb-5 gap-5 md:items-stretch">
      <div class="basis-1/3">
        <div class>
          <div class="md:sticky md:top-5">
            <h1>
              Title
            </h1>
            <span class="hidden md:block">
              <a class="border-2 border-slate-500 uppercase font-bold bg-white dark:bg-slate-900 px-4 py-2 text-slate-800 dark:text-slate-200 hover:bg-slate-100 dark:hover:bg-slate-800 hover:text-black dark:hover:text-white"
                 href="/"
              >
                « Home
              </a>
            </span>
          </div>
        </div>
      </div>
      <div class="w-full">
        <p>
          Causa sumo magni eaque timor consequuntur cruentus.
        </p>
        <p>
          Suus uterque arto aspicio paulatim votum uredo consectetur. Trepide ager aedificium cursus abscido bardus valens defendo vere infit.
        </p>
        <p>
          Aliquid acsi clibanus totidem. Carbo modi adstringo defluo aegre tardus demitto talis dolor administratio. Suggero cattus tumultus creber asperiores timor.
        </p>
        <p>
          Tondeo cito stipes cohibeo vulgivagus sono cultellus amoveo auditor. Depopulo valens desipio subito culpa decerno attollo debitis. Teneo pecco maxime. Ago uxor nostrum nulla tot.
        </p>
        <p>
          Conduco astrum vestigium adstringo tandem aro. Confido tantillus decet ager stella unde basium. Praesentium verbera vesper aperiam ager aggredior deorsum aptus sublime. Aggredior curriculum ustulo libero damnatio annus. Baiulus derelinquo tondeo correptius delibero vesco.
        </p>
      </div>
    </div>
  </div>
</div>
`;

exports[`Layout/Article NoImageNoTitle smoke-test 1`] = `
<div class="bg-slate-100 dark:bg-slate-900 text-slate-900 dark:text-slate-300">
  <div class="container mx-auto py-5 px-3">
    <div class="flex flex-col md:flex-row mb-5 gap-5 md:items-stretch">
      <div class="basis-1/3">
      </div>
      <div class="w-full">
        <p>
          Celer cras studio sordeo triumphus adamo omnis repellat.
        </p>
        <p>
          Adduco speculum tamdiu cruciamentum demergo subvenio. Vociferor illo alioqui.
        </p>
        <p>
          Adsum solitudo odio. Tantillus tenus dolorem natus verumtamen ambulo cognomen subseco depromo solvo. Admiratio commodo textor auditor earum cogo cursim coniecto tertius spoliatio.
        </p>
        <p>
          Adnuo denique timidus attonbitus pax commemoro. Tutis tametsi audentia facilis cur amo vox derelinquo. At veritatis stella versus. Aliqua quibusdam delectus comburo desipio deprimo combibo.
        </p>
        <p>
          Adiuvo apparatus carbo animus allatus decumbo capitulus adhaero timor. Vehemens cunabula adimpleo corroboro cogito. Vomer tonsor fugit eum quidem. Abscido angulus deorsum terreo voveo dapifer corrupti. Alo attonbitus deinde.
        </p>
      </div>
    </div>
  </div>
</div>
`;

exports[`Layout/Article Offset smoke-test 1`] = `
<div class="bg-slate-100 dark:bg-slate-900 text-slate-900 dark:text-slate-300">
  <div class="container mx-auto py-5 px-3">
    <div class="flex flex-col md:flex-row mb-5 gap-5 md:items-stretch">
      <div class="basis-1/3">
        <div class="-mt-10 md:-mt-20 h-full">
          <img alt
               width="500"
               height="500"
               decoding="async"
               data-nimg="1"
               class="border-2 border-slate-500 mb-5 bg-white"
               srcset="https://picsum.photos/seed/Qa5jLTfb/640/480?w=640&amp;q=75 1x, https://picsum.photos/seed/Qa5jLTfb/640/480?w=1080&amp;q=75 2x"
               src="https://picsum.photos/seed/Qa5jLTfb/640/480?w=1080&amp;q=75"
               style="color: transparent;"
          >
          <div class="md:sticky md:top-5">
            <h1>
              Title
            </h1>
            <span class="hidden md:block">
              <a class="border-2 border-slate-500 uppercase font-bold bg-white dark:bg-slate-900 px-4 py-2 text-slate-800 dark:text-slate-200 hover:bg-slate-100 dark:hover:bg-slate-800 hover:text-black dark:hover:text-white"
                 href="/"
              >
                « Home
              </a>
            </span>
          </div>
        </div>
      </div>
      <div class="w-full">
        <p>
          Depulso mollitia benevolentia allatus.
        </p>
        <p>
          Vae artificiose quis verto supra vulgus acceptus. Paens collum iusto dignissimos articulus cicuta.
        </p>
        <p>
          Tendo sub amitto comis. Stultus verbera iste blanditiis creta conventus vespillo. Turba sum contabesco crudelis adimpleo campana cogo crastinus cubicularis.
        </p>
        <p>
          Sufficio iusto vulnero abeo constans reprehenderit tempus. Optio adicio stultus. Arcus clarus cubitum tenax caput corona coniecto uterque dolores adamo. Ipsam amiculum caput aperte vado comminor degero solvo.
        </p>
        <p>
          Tyrannus claudeo ipsum toties. Laboriosam creo spiritus comes tolero considero ustilo assumenda cura. Adsum confido demulceo ipsam eligendi nesciunt nobis enim colligo. Verto voveo sollers. Conscendo beatus hic clarus deduco eius cimentarius veniam clementia vorago.
        </p>
      </div>
    </div>
  </div>
</div>
`;
